servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/DefaultCompositeCloseable.java

```diff
@@ -112,39 +125,85 @@ public void closeGracefully() {
+        resetState();
+    }
+
+    private void resetState() {
+        closeAsync = closeAsyncGracefully = null;
+    }
+
+    private Operand getOrAddMergeOperand() {
+        return getOrAddOperand(true, true);
+    }
+
+    private Operand getOrAddConcatOperand(boolean append) {
+        return getOrAddOperand(append, false);
+    }
+
+    private Operand getOrAddOperand(boolean append, boolean isMerge) {
+        final Operand operand;
+        if (operands.isEmpty()) {
+            operand = new Operand(isMerge);
+            operands.add(operand);
+        } else {
+            Operand rawOperand = append ? operands.getLast() : operands.getFirst();
+            if (isMerge == rawOperand.isMerge) {
+                operand = rawOperand;
+            } else {
+                operand = new Operand(isMerge);
+                if (append) {
+                    operands.add(operand);
+                } else {
+                    operands.addFirst(operand);
+                }
+            }
+        }
+        return operand;
     }
 
     private void concatCloseableDelayError(final AsyncCloseable closeable) {
-        closeAsync = closeAsync.concat(closeable.closeAsync().onErrorComplete(th -> {
-            //TODO: This should use concatDelayError when available.
-            LOGGER.debug("Ignored failure to close {}.", closeable, th);
-            return true;
-        }));
-        closeAsyncGracefully = closeAsyncGracefully.concat(closeable.closeAsyncGracefully().onErrorComplete(th -> {
-            //TODO: This should use concatDelayError when available.
-            LOGGER.debug("Ignored failure to close {}.", closeable, th);
-            return true;
-        }));
+        final Operand operand = getOrAddConcatOperand(true);
+        operand.closables.add(closeable);
+        resetState();
     }
 
     private void prependCloseableDelayError(final AsyncCloseable closeable) {
-        closeAsync = closeable.closeAsync().onErrorComplete(th -> {
-            //TODO: This should use prependDelayError when available.
-            LOGGER.debug("Ignored failure to close {}.", closeable, th);
-            return true;
-        }).concat(closeAsync);
-        closeAsyncGracefully = closeable.closeAsyncGracefully().onErrorComplete(th -> {
-            //TODO: This should use prependDelayError when available.
-            LOGGER.debug("Ignored failure to close {}.", closeable, th);
-            return true;
-        }).concat(closeAsyncGracefully);
+        final Operand operand = getOrAddConcatOperand(false);
+        operand.closables.add(closeable);
+        resetState();
+    }
+
+    private Completable buildCompletable(Function<AsyncCloseable, Completable> closerFunc) {
+        Completable result = completed();
+        for (Operand operand : operands) {
+            if (operand.isMerge) {
+                result = result.mergeDelayError(operand.closables.stream().map(closerFunc).toArray(Completable[]::new));
```

Comment:
Any pros/cons of using `completed().mergeDelayError(array)` vs static `mergeAllDelayError(array)`? If there are any, have you considered to add a static `concatAll(array)` utility instead of `Completable.concat(array)`?

Reply:
`concatAll` may have a small advantage if you just have an array bcz no need to bootstrap with `completed()` (initialization of subscribe a bit more streamlined) however the control flow here mixes the two operations so it makes it cleaner to just compose each operator individually.